packages/eslint-plugin-react-hooks/__tests__/ESLintRulesOfHooks-test.js

```diff
@@ -479,6 +479,55 @@ const tests = {
         }
       `,
     },
+    {
+      // Nested component declarations are bad but should be covered by other ESLint rules
+      code: normalizeIndent`
+        function App() {
+          return createElement(Child, {
+            Component: () => {
+              const [myState, setMyState] = useState(null);
+            },
+          });
+        }
+      `,
+    },
+    {
+      // Nested component declarations are bad but should be covered by other ESLint rules
+      code: normalizeIndent`
+        function JSXApp() {
+          return (
+            <Child
+              Component={() => {
+                const [myState, setMyState] = useState(null);
+              }}
+            />
+          );
+        }
+      `,
+    },
+    {
+      only: true,
+      // because
+      // const useNamed = async () => useQuery();
+      // <Foo useData={useNamed} />;
+      // would also be valid
+      code: normalizeIndent`
+        function useQuery() {
+          return React.useState(null);
+        }
+        function App() {
+          const useNamed = async () => useQuery();
+
+          return (
+            <div className="App">
+              <Foo useData={async () => useQuery()} />
+              <Foo useData={useNamed} />
```

Comment:
This is already accepted on `main` but not `<Foo useData={async () => useQuery()} />`. Test taken from https://github.com/facebook/react/issues/23230